### PR TITLE
Query stream issue #151

### PR DESCRIPTION
**Pull request by RTrackerDev**
Minor fix for https://github.com/objectbox/objectbox-dart/issues/151

### CODE REVIEW

**Comment by RTrackerDev** [comment]
@greenrobot @Buggaboo @vaind could you check please?

**Review by vaind — APPROVED** [review]
Looks reasonable to me though I'd be happier if @Buggaboo could have a quick look as an author of the code. Unfortunately, I'm not 100% confident with this section yet.

**Comment by vaind** [comment]
And almost forgotten... is it possible to test this? Would a small test case to verify the issue was there before (and is not anymore after the changes) be too hard to do? That'd help make avoid future regressions.

**Comment by Buggaboo** [comment]
```dart
test(
      'Only observers of a single entity are notified, no cross-entity observer notification',
      () async {
    // setup listeners
    final box2 = Box<TestEntity2>(env.store);

    var counter1 = 0, counter2 = 0;

    final query2 = box2.query().build();
    final queryStream2 = query2.findStream();
    final subscription2 = queryStream2.listen((_) {
      counter2++;
    });

    final query1 = box.query().build();
    final queryStream1 = query1.findStream();
    final subscription1 = queryStream1.listen((_) {
      counter1++;
    });

    // counter2 test #1
    final t2 = TestEntity2();
    box2.put(t2);

    await Future.delayed(Duration(seconds: 0));
    expect(counter1, 0);
    expect(counter2, 1);

    // counter1 test #1
    final t1 = TestEntity();
    box.put(t1);

    await Future.delayed(Duration(seconds: 0));
    expect(counter1, 1);
    expect(counter2, 1);

    // counter1 many test #2
    final ts1 = [1, 2, 3].map((i) => TestEntity(tInt: i)).toList();
    box.putMany(ts1);

    await Future.delayed(Duration(seconds: 0));
    expect(counter1, 2);
    expect(counter2, 1);

    // counter2 many test #2
    final ts2 = [1, 2, 3].map((i) => TestEntity2()).toList();
    box2.putMany(ts2);

    await Future.delayed(Duration(seconds: 0));
    expect(counter1, 2);
    expect(counter2, 2);

    query1.close();
    query2.close();

    await subscription1.cancel();
    await subscription2.cancel();
  });
```

I'm looking into it.

**Comment by RTrackerDev** [comment]
> And almost forgotten... is it possible to test this? Would a small test case to verify the issue was there before (and is not anymore after the changes) be too hard to do? That'd help make avoid future regressions.

I wanted to add tests but I have problems running on mac OS

**Comment by Buggaboo** [comment]
Have you run 'install.sh'?

File: lib/src/observable.dart [hunk 1]
```diff
@@ -9,7 +9,7 @@ import 'util.dart';
 // ignore_for_file: non_constant_identifier_names
 
 // dart callback signature
-typedef Any = void Function(Pointer<Void>, Pointer<Uint32>, int);
```
Comment (Buggaboo) [inline lib/src/observable.dart]: The change to the 2nd param is [questionable](https://github.com/objectbox/objectbox-c/blob/master/include/objectbox.h#L1543-L1546).

[It's an array of Uint32s](https://github.com/objectbox/objectbox-c/blob/master/include/objectbox.h#L56).

Comment (RTrackerDev) [reply lib/src/observable.dart]: > The change to the 2nd param is [questionable](https://github.com/objectbox/objectbox-c/blob/master/include/objectbox.h#L1543-L1546).
> 
> [It's an array of Uint32s](https://github.com/objectbox/objectbox-c/blob/master/include/objectbox.h#L56).
> 
> The array should be unpacked, and the entityId it contains should be used to trigger the specific callback(s) indexed by storePtr and entityId. I might have been a bit sloppy.

@Buggaboo I didn't change it.

**Comment by RTrackerDev** [comment]
> Have you run 'install.sh'?
> 
> Also stripped the 'dart' exec?

install.sh ?

**Comment by RTrackerDev** [comment]
> Have you run 'install.sh'?
> 
> Also stripped the 'dart' exec?

Thank you i resolved it

File: lib/src/observable.dart (hunk 1 above)
Comment (Buggaboo) [reply lib/src/observable.dart]: Sure. I'm on it.

**Comment by Buggaboo** [comment]
I have this feeling the callback in C is not behaving the way it's documented.

**Comment by RTrackerDev** [comment]
@vaind @Buggaboo I added unit test and resolve conflicts

**Comment by RTrackerDev** [comment]
> I have this feeling the callback in C is not behaving the way it's documented.

He is behaving correctly. The problem is that everyone subscribes to a single stream controller, which notifies everyone.

**Comment by RTrackerDev** [comment]
@Buggaboo In my opinion, the best solution is to move this StreamController as value to the query and when the request is closed, close the controller

**Comment by Buggaboo** [comment]
@vaind Do we still have to close Queries? Or did I forgot to add them?

**Comment by Buggaboo** [comment]
I found another (conceptual) bug. Fixing it.

**Comment by RTrackerDev** [comment]
> I found another (conceptual) bug. Fixing it.
> 
> The number of C callback notifications should not increase linearly along with the number of dart listeners. Only one C observer per store should be initialized to serve notifications.

@vaind @Buggaboo But this PR you could merge.

**Comment by Buggaboo** [comment]
I took your changes with a fix for another bug I found. I have to redo it again because I based off the wrong branch.

**Comment by Buggaboo** [comment]
> > I found another (conceptual) bug. Fixing it.
> > The number of C callback notifications should not increase linearly along with the number of dart listeners. Only one C observer per store should be initialized to serve notifications.
> 
> @vaind @Buggaboo But this PR you could merge.

I was looking at an old branch :(, that other problem was fixed already.

**Comment by RTrackerDev** [comment]
> > > I found another (conceptual) bug. Fixing it.
> > > The number of C callback notifications should not increase linearly along with the number of dart listeners. Only one C observer per store should be initialized to serve notifications.
> > 
> > 
> > @vaind @Buggaboo But this PR you could merge.
> 
> I was looking at an old branch :(, that other problem was fixed already.

What problem was fixed?

**Comment by Buggaboo** [comment]
> > > > I found another (conceptual) bug. Fixing it.
> > > > The number of C callback notifications should not increase linearly along with the number of dart listeners. Only one C observer per store should be initialized to serve notifications.
> > > 
> > > 
> > > @vaind @Buggaboo But this PR you could merge.
> > 
> > 
> > I was looking at an old branch :(, that other problem was fixed already.
> 
> What problem was fixed?

Only one C observer per store should be initialized to serve notifications.

**Comment by Buggaboo** [comment]
I sent @RTrackerDev a PR with an extra test and a type check appeaser. Once that's merged to this branch, this one is ready to go.

**Review by Buggaboo — APPROVED** [review]
My changes weren't really helpful. In hindsight. This is good to go.

**Comment by RTrackerDev** [comment]
@vaind @Buggaboo Could you merge if its Ok?

**Comment by Buggaboo** [comment]
@vaind could do it. I don't have the permissions. @RTrackerDev he's a busy guy. Regardless, your contribution is appreciated. (At least by me)

**Comment by vaind** [comment]
thx for the fix @RTrackerDev and the support @Buggaboo, merging. 

FYI there will be major changes to query streams implementation, to support asynchronous callbacks (across multiple isolates), see #145.

**Pull request merged**